projects/RabbitMQ.Client/client/ClientArrayPool.cs

```diff
@@ -0,0 +1,114 @@
+﻿// This source code is dual-licensed under the Apache License, version
+// 2.0, and the Mozilla Public License, version 2.0.
+//
+// The APL v2.0:
+//
+//---------------------------------------------------------------------------
+//   Copyright (c) 2007-2020 VMware, Inc.
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       https://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//---------------------------------------------------------------------------
+//
+// The MPL v2.0:
+//
+//---------------------------------------------------------------------------
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+//
+//  Copyright (c) 2007-2020 VMware, Inc.  All rights reserved.
+//---------------------------------------------------------------------------
+
+using System;
+using System.Buffers;
+using System.Collections.Concurrent;
+using System.Runtime.CompilerServices;
+
+namespace RabbitMQ.Client
+{
+    internal static class ClientArrayPool
+    {
+        private static readonly ConcurrentDictionary<byte[], bool> _checkouts;
+
+        private static readonly bool s_useArrayPool = true;
+        private static readonly bool s_trackCheckouts = false;
+
+        static ClientArrayPool()
+        {
+            if (false == bool.TryParse(
+                Environment.GetEnvironmentVariable("RABBITMQ_CLIENT_USE_ARRAY_POOL"),
+                out s_useArrayPool))
+            {
+                s_useArrayPool = true;
+            }
+
+            if (bool.TryParse(
+                Environment.GetEnvironmentVariable("RABBITMQ_CLIENT_TRACK_ARRAY_POOL_CHECKOUTS"),
+                out s_trackCheckouts))
+            {
+                if (s_trackCheckouts)
+                {
+                    _checkouts = new ConcurrentDictionary<byte[], bool>();
+                }
+            }
```

Comment:
Consider adding [`AppContext`](https://learn.microsoft.com/en-us/dotnet/api/system.appcontext) support for these settings.

They can be set:
- Per application/process
  - Can be set at run-time (provided it is set before use).
  - Can be set at compile/build time using **MSBuild** properties.

Here's an example usage [from .NET](https://source.dot.net/#System.Net.Http/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.NtAuth.cs,48):

```csharp
        private static bool UsePortInSpn
        {
            get
            {
                int usePortInSpn = s_usePortInSpn;
                if (usePortInSpn != -1)
                {
                    return usePortInSpn != 0;
                }
 
                // First check for the AppContext switch, giving it priority over the environment variable.
                if (AppContext.TryGetSwitch(UsePortInSpnCtxSwitch, out bool value))
                {
                    s_usePortInSpn = value ? 1 : 0;
                }
                else
                {
                    // AppContext switch wasn't used. Check the environment variable.
                   s_usePortInSpn =
                       Environment.GetEnvironmentVariable(UsePortInSpnEnvironmentVariable) is string envVar &&
                       (envVar == "1" || envVar.Equals("true", StringComparison.OrdinalIgnoreCase)) ? 1 : 0;
                }
 
                return s_usePortInSpn != 0;
            }
        }
```

And [another](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs,33):

```csharp
        private static bool TryGetStringValue(string switchName, string envVariable, [NotNullWhen(true)] out string? value)
        {
            value = AppContext.GetData(switchName) as string;
            if (string.IsNullOrEmpty(value))
            {
                value = Environment.GetEnvironmentVariable(envVariable);
                if (string.IsNullOrEmpty(value))
                {
                    return false;
                }
            }
 
            return true;
        }
```

But this should, probably, be a separate PR.

Reply:
I'm still confused why this whole class is needed. My assumption is it was added for tracking purposes. See my comment in https://github.com/rabbitmq/rabbitmq-dotnet-client/pull/1347#discussion_r1396380470

ArrayPool has already event sources that allow to trace rent and return calls.

Reply:
Yep, see #1421.